cmd/copy.go

```diff
@@ -17,79 +16,141 @@ import (
 var getSessionIDRegex = regexp.MustCompile(`^sess:([^/]+)`)
 
 func Copy(cmd *cobra.Command, args []string) error {
-	scpArgs := make([]string, 0, len(args))
-	var targetExec *types.Exec
+	exec, err := getTargetExec(cmd, args)
+	if err != nil {
+		return err
+	}
+
+	if exec == nil {
+		return fmt.Errorf("At least one remote host must be specified")
+	}
+	if exec.Connection == nil {
+		return fmt.Errorf("Target session must have an active connection")
+	}
+	if exec.SSHKey.PublicKey == nil && config.SSHPublicKeyPath == "" {
+		return fmt.Errorf("Failed to identify public key, check your Unweave config file or specify it manually")
+	}
+
+	scpArgs, err := formatPaths(exec, args)
+	if err != nil {
+		ui.Infof("❌ Unsuccessful copy: %s", err.Error())
+	}
+	ui.Infof(fmt.Sprintf("🔄 Copying %s => %s", scpArgs[0], scpArgs[1]))
+
+	switch {
+	case isLocalDirToRemoteCopy(args):
+		err = copySourceAndUnzip(exec.ID, scpArgs[0], splitSessFromDirpath(args[1]), *exec.Connection, *exec.SSHKey.PublicKey)
+	default:
+		err = copySourceSCP(scpArgs[0], scpArgs[1], *exec.SSHKey.PublicKey)
+	}
+
+	if err != nil {
+		ui.Infof("❌ Unsuccessful copy %s => %s", scpArgs[0], scpArgs[1])
+		return nil
+	}
 
+	ui.Infof("✅  Copied %s => %s", scpArgs[0], scpArgs[1])
+	return nil
+}
+
+func isLocalDirToRemoteCopy(args []string) bool {
+	if strings.Contains(args[0], "sess:") {
+		return false
+	}
+
+	pathInfo, err := os.Stat(args[0])
+	if err != nil || pathInfo == nil {
+		return false
+	}
+
+	return pathInfo.IsDir()
+}
+
+func isRemoteDirToLocalCopy(args []string) bool {
+	if strings.Contains(args[1], "sess:") {
+		return false
+	}
+
+	pathInfo, err := os.Stat(args[1])
```

Comment:
Same issue as above accessing slice elements. This it prone to panics. The args parsing should happen in the cobra handler entry points.